### ISSUE TITLE
Swarm Memory / Cpu / Network usage sky rockets after 220 hosts

### ISSUE THREAD

**Issue by anandkumarpatel** [docker/classicswarm#2540]
Hi I am seeing swarm manager's cpu, ram, and network usage skyrocket after ~220 hosts
Interesting things to notes: 

1. Network traffic *into* swarm manager is increasing 
2. Number of containers across the system it constant
3. Memory and cpu usage increase at a faster rate after 220 hosts
4. Number of api calls going to swarm manager has been constant

Has anyone seen this before and is it expected?

Swarm agent version: 1.2.0
Swarm manager version: 1.2.3
all docker hosts are docker 1.12.0

![screen shot 2016-11-07 at 9 58 41 am](https://cloud.githubusercontent.com/assets/2194285/20069956/7d4cf8fa-a4d3-11e6-8122-7b9596e9798d.png)

**Comment by anandkumarpatel** [comment]
Another note: I noticed all of that network traffic is coming from our consul cluster. Metrics show both Swarm and Consul receive 7/mb a second.

**Comment by dongluochen** [comment]
@anandkumarpatel Thanks for reporting this problem! 7MB/s seems like a lot of data. Does node membership keep flapping in Consul? Can you enable logging on manager (`swarm -l debug manage ...`) and send us the logs?

**Comment by anandkumarpatel** [comment]
@dongluochen Thanks for the response!! here are the logs 
https://gist.github.com/anandkumarpatel/11c5926fcd59d8f6be335cbe08c1814c

What do I look for to see node flapping? 🐦

**Comment by dongluochen** [comment]
@anandkumarpatel Thanks for the logs. Consul shows 408 nodes consistently so they are not flapping (going in and out of Consul). 

```
time="2016-11-19T03:59:59Z" level=debug msg="Watch triggered with 408 nodes" discovery=consul
```

When the primary manager (under debug mode) is busy, can you dump the goroutines using the following command and share it with us? 
```
wget -O goroutine http://primary:port/debug/pprof/goroutine?debug=1
```

**Comment by anandkumarpatel** [comment]
@dongluochen  goroutines: 
https://gist.github.com/anandkumarpatel/29d37ebd3ee82942b6e7d17aa95e1888

>>> Network into swarm manager is increasing
>>>>>>Also, how many networks do you create in the swarm?

we do not create any docker networks in swarm. Updated description to say Network traffic into swarm. When I did `iptarf` and `iftop` it showed most traffic was coming from consul server.

more info: 
`docker network ls`
https://gist.github.com/anandkumarpatel/867aecd5e98f1cffc7519461a163da46


`docker info`
```
Containers: 499827
 Running: 2891
 Paused: 0
 Stopped: 496936
Images: 7497
Server Version: swarm/1.2.3
Role: primary
Strategy: spread
Filters: health, port, containerslots, dependency, affinity, constraint
Nodes: 438
...
to many nodes
...
Plugins:
 Volume:
 Network:
Swarm:
 NodeID:
 Is Manager: false
 Node Address:
Security Options:
Kernel Version: 3.13.0-85-generic
Operating System: linux
Architecture: amd64
CPUs: 1030
Total Memory: 4.123 TiB
Name: f3f899cacf13
Docker Root Dir:
Debug Mode (client): false
Debug Mode (server): false
WARNING: No kernel memory limit support
```

**Comment by anandkumarpatel** [comment]
we reached 9gb at 445 hosts today

**Comment by dongluochen** [comment]
@anandkumarpatel The graphs are very helpful. Is the network IO per machine, or aggregated total? Does your workload contribute to the network IO? 

> we reached 9gb at 445 hosts today 

At around 200 nodes, memory was at 1GB, while at 445, it's 9GB. It's not a linear relationship. I suspect there are memory leakage. Do you see memory usage continuously increase when the number of hosts is stable?

**Comment by anandkumarpatel** [comment]
@dongluochen 
>>> Is the network IO per machine, or aggregated total

there are 2 network graphs. 
1. `Swarm Network IO (received and -send)`: This is the network traffic inside of our single swarm manager instance.
2. `Console network`: Traffic in and out of our 3 consul servers, one color per server.

>>> Does your workload contribute to the network IO?

No, our workload has little to no network traffic.

>>>I suspect there are memory leakage.

The problem with this theory is that after I redeploy swarm, memory and cpu usage return to previous levels right away. A redeploy means we create a new container from the swarm image and direct *all* traffic to the new manager.

>>> Do you see memory usage continuously increase when the number of hosts is stable?
I zoomed into a period where we have 347 hosts for 12 hours. Memory stated ~5.53gb and ended ~5.65
![screen shot 2016-11-22 at 6 42 03 pm](https://cloud.githubusercontent.com/assets/2194285/20549690/86d53f10-b0e3-11e6-861d-e4dbd516b275.png)

The goroutines dump showed 4001 number of goroutines. We are using around 8gb does that mean 2mb per goroutine?

**Comment by dongluochen** [comment]
> The goroutines dump showed 4001 number of goroutines. We are using around 8gb does that mean 2mb per goroutine?

The goroutine in Swarm shouldn't take that much memory. I suspect there are some resource not properly released.

**Comment by anandkumarpatel** [comment]
I found the issue and wrote a blog post on it: 
https://runnable.com/blog/why-is-docker-swarm-consuming-all-available-ram

TLDR;
We had 500k *stopped* containers which we did not have on our graphs above. Since swarm stores inspect info for all containers across all nodes in memory and we pack ~12kb of data per inspect object using labels, we used a lot of ram!

**Comment by dongluochen** [comment]
Thanks Anand. I replied to your linkedin post https://www.linkedin.com/pulse/why-docker-swarm-consuming-all-available-ram-anand-patel

`
... Swarm (sometimes people also call it classic swarm, to differentiate 
from swarmkit) manager keeps track of all the containers in the cluster so it can decide resource 
management. Containers not in running state are kept because users may do 2-step deployment. In 
the first step they create all the containers which downloads images and creates the containers on 
all the nodes. In the second step they start the containers which has real impact to service. Because 
all the preparation work has been done in step 1, this can be done fast.
`

2-step deployment is common practice. That's why swarm manager keeps all the containers, disregard their status.
